datafusion/src/physical_plan/coercion_rule/binary_rule.rs

```diff
@@ -77,7 +77,7 @@ pub(crate) fn coerce_types(
 }
 
 fn comparison_eq_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
-    if lhs_type == rhs_type {
+    if lhs_type == rhs_type && !is_dictionary(lhs_type) {
```

Comment:
the arrow `eq` kernels don't support `DictionaryArray` <==> `DictionaryArray` comparisons yet, so we need to handle this case specially. I will also file a ticket to add native DictionaryArray comparison support.

Reply:
https://github.com/apache/arrow-rs/issues/1201